crates/bevy_time/src/fixed.rs

```diff
@@ -0,0 +1,320 @@
+use bevy_ecs::world::World;
+use bevy_reflect::Reflect;
+use bevy_utils::Duration;
+
+use crate::{time::Time, virt::Virtual, FixedUpdate};
+
+/// The fixed timestep game clock following virtual time.
+///
+/// It specialization of the [`Time`] structure. **For method documentation, see
+/// [`Time<Fixed>#impl-Time<Fixed>`].**
+///     
+/// It is automatically inserted as a resource by
+/// [`TimePlugin`](crate::TimePlugin) and updated based on
+/// [`Time<Virtual>`](Virtual). The fixed clock is automatically set as the
+/// generic [`Time`] resource during [`FixedUpdate`](bevy_app::FixedUpdate)
+/// schedule processing.
+///
+/// The fixed timestep game clock runs at a regular interval and is extremely
+/// useful for steady, frame-rate independent gameplay logic and physics.
+///
+/// The default [`timestep()`](Time::timestep) is 64 hertz, or 15625
+/// microseconds. The reason this value is chosen is that it is relatively close
+/// to a standard monitor refresh rate of 60 hertz, but is accurately
+/// representable as a 32-bit and 64-bit floating point number.
```

Comment:
It's not related to framerate. 64Hz was chosen because it's a power-of-two, which means `time.delta_seconds()` will be an exact value instead of an approximation. That makes doing math in `FixedUpdate` more accurate. 

```rust
position += velocity * time.delta_seconds();
```

If you integrate time like this and `time.delta_seconds()` is 15.625ms, rounding errors are minimized.

There's no benefit to matching the timestep to a common monitor refresh rate. That doesn't help things look smoother.

Reply:
The problem with 60 Hz is that it may become a pathological case of running an alternating sequence of 0 and 2 fixed updates.

There would be a point in making the fixed timestep be something like 4x the framerate, so 256 Hz (or 240 Hz if not caring about division accuracy). That way if some frames have 3 fixed updates and some have 5, the results will not be as jarring as with zero and 2. But that might be a worse default as it's not usually done by engines that way.

But ultimately – it's just some number as some number has to be picked as the default. Unity default is 50 Hz for the fixed update.

Reply:
60Hz is a nice default value because no one will question it. It's the defacto minimum frame rate, and every player device is expected to handle running logic 60 times a second. 2x and 4x? Not so much.

But 60Hz is just a convention. There's no benefit to having the tick rate match the average FPS. The number of ticks per frame will actually oscillate the most when that happens.

The thing for users to recognize is you can't get smooth rendering reliably without interpolation (or extrapolation), and having that will make any tick rate look smooth. So frame and tick rates are independent, games have just settled on "stuff should run at least 60 times a second" to reduce input lag.

64Hz is just the closest power-of-two to 60Hz, but it's also pretty common (e.g. Counter Strike). And again, the reason to use a power-of-two as the default is they don't cause rounding errors.

Reply:
> ```rust
> position += velocity * time.delta_seconds();
> ```
> 
> If you integrate time like this and `time.delta_seconds()` is 15.625ms (64Hz), rounding errors are minimized.

That is correct, but even if the value is not `power-of-two`, the rounding error will be within half of [ULP](https://en.wikipedia.org/wiki/Unit_in_the_last_place). That is not 0 but quite small.

> There's no benefit to matching the timestep to a common monitor refresh rate. That doesn't help things look smoother.

That is true without VSync or when #9034 is merged, but not with VSync. With VSync, two ticks are run a frame, which causes periodic game speed jumps. Also, the maximum CPU usage is limited to half to prevent lag spikes.

The disadvantages outweigh the advantages of using 64hz. Unless #9034 has been merged.

Well, but if #9034 is merged I will set the default value to 125hz. This value is consistent with the polling rate for typical input devices, thus optimizing input responsiveness.

> The problem with 60 Hz is that it may become a pathological case of running an alternating sequence of 0 and 2 fixed updates.

I think this can be avoided by buffering the updates. If 2 updates are needed, 1 update and 1 buffer. If 3 updates are needed, update 2 and buffer 1. If 0 updates are needed, then update if buffered.

Reply:
> The disadvantages outweigh the advantages of using 64hz. Unless #9034 has been merged.

Just because some users can get lucky by enabling VSync on a 60Hz monitor, that's not a good reason to set the tick rate at 60Hz. That isn't an endorsable solution.

The closest thing to an official position has always been "interpolation or sub-stepping are the only robust ways to render fixed stuff smoothly".

> I will set the default value to 125hz. This value is consistent with the polling rate for typical input devices, thus optimizing input responsiveness.

You can do as you like, but we are going
in the direction of decoupling the `winit` event loop from the app updates.

Input events cannot be observed in realtime because the `winit` loop is blocked when the app updates, but that'll change with #9122. Once we merge #9122, it'll be possible for `FixedUpdate` to receive input events in a framerate-independent way.

The frame rate, tick rate, and input "polling" rate will all be decoupled.

Reply:
Even if fixed timestep is 60 Hz, and display refresh is 60 Hz, it will not mean that there will always be exactly one fixed update per frame. Display clock frequency never matches system clock frequency exactly – no two clocks ever do – so they will diverge if the game runs against the system clock.

The only way to avoid this is to tie the game `elapsed` and `delta` etc. to the display refresh clock. But we don't want to do that (even if VSync is enabled), because display clocks vary (59.997 Hz and 59.940 Hz being explicit examples of intentionally not being 60 Hz) and an important point of fixed timestep is to ensure the game behaves consistently regardless of frame rate - and regardless of the computer the game is run on.

As for input events... IMHO, input events should be tagged with `Instant`, which then needs to be translated to virtual game time (by scaling it between last update and this update, can discuss this more why I believe this is right), and then made visible to the game if the virtual game time is past the instant of the input (in addition exposing the virtual timestap of the input event directly to the game as well, if the game wants to do sub-frame accuracy for inputs). Frame rate should not enter into it in any way. If there's a need to poll inputs because they don't arrive as an event stream natively, then this polling loop should have nothing to do with the game loop, as it's only purpose is to fetch inputs, timestamp them with instant and put them in a queue.

And as for doing ticks where nothing is rendered... How I see it is that the only meaningful output of a game is when it displays something (not true for networked games, dedicated servers, etc.), and the point of running systems is to compute what needs to be shown. For "perfect" games, just running Update schedule will run any amount of game logic with exact results – but most games are not "perfect", but instead use a fixed timestep to ensure their logic is run in discrete chunks regardless of the rate of updates. Running any of this update logic without rendering serves no direct purpose, as only the outputs matter – the only benefit being that if calculating the next state takes long, then the more that can be calculated before rendering the better because it means the actual render pass will be shorter which means there will be less latency from "we can start rendering a frame" to "frame is rendered". Concretely this would mean that as soon as it is clear that no more inputs will be arriving that would affect the next fixed update step, the next fixed update step should be calculated. However, there should be exactly zero behaviour impact from this – the only impact being that we might use the time between frames more efficiently by splitting up the computation, reducing latency for the actual frame rendering. This would be a reduction of latency of probably less than 10 milliseconds in real world, so not sure if it's worth the complexity.

All that said though, I'd be happy to put 125 Hz as the fixed update timestep. The only reason why I didn't suggest a "much higher than refresh" value to start with is that Bevy used to have 60 Hz as the fixed timestep, and Unity has 50 Hz as the fixed timestep, so doubling the default rate (or more) is probably disruptive for a bunch of existing games and a cause of surprise for new users. But if someone else makes that judgement call, I'm not complaining as I think that's a better default ultimately.

Reply:
And separately... buffering is probably just weird. If the refresh Hz is above fixed timestep, it will be just normal behaviour of no buffering – basically a sequence of `0, 1, 1, 1, 1 ... 1, 0, 1, 1, 1, 1 ...` over and over again. If the refresh Hz is below fixed timestep (like 60 Hz vs 64 Hz), then it will go to a sequence of `1, 1, 1, 1 ... 1, 2, 1, 1, 1, 1 ...` with the 2 updates being run if *three* updates would really be queued. It does smooth out bumps in the framerate... but this makes the time run on average with 1.5 frame lag compared to virtual game time, which is silly but also really problematic as the whole point of this exercise is to ensure that time inside fixed timestep is never desynced from time outside fixed timestep. If we ever "buffer" frames, be it via "max ticks per update" or some other method, then `Time<Fixed>` cannot be trusted to be between the previous `Time<Virtual>` and the next `Time<Virtual>`. And this is an approach that I do not agree with, so it means someone else has to finish the pull.

Reply:
> input events should be tagged with an `Instant`
> that is then be translated into virtual game time
> and then the event should be made visible to the game if the virtual time is past that time

You don't need to convince the maintainers. This has been the goal for a long time, but it's blocked on merging a solution like #9122. If we tried to timestamp inputs right now, they'd all just be aliased to the `Instant` a frame starts.

> 125 Hz

I'm happy to volunteer to make the judgement call. 

100+Hz is too much. Games normally target 60[^1] ticks/sec. That's the minimum players expect today, and it's still what most games aim for in 2023. If the default is set significantly higher than 60Hz, too many users will have to lower it for their games to have acceptable performance. 

A default is supposed to be serve most users so they don't need to change it, not show off how good something can be in especially well-optimized cases. So 60Hz is the minimum value[^2], and raising it higher makes it worse as a default.

64Hz hits a sweet spot of being barely above 60Hz while giving some practical accuracy benefits. So, with all that said, 60Hz and 64Hz are the only acceptable choices. And if it'll end the debate, just go with 60 lol. 

[^1]: 60Hz does come from the lowest common denominator in screen refresh rates, but setting the minimum expectation is as far as that relationship goes.